### ISSUE TITLE
App might crash when calling into Castle from non-main threads

### ISSUE THREAD

**Issue by tiwoc** [castle/castle-ios#58]
When calling any Castle methods from non-main threads, the app could crash because the Highwind library proceeds to call UIKit APIs synchronously that should only be called from the main thread. In some cases, this only leads to runtime warnings about API misuse in Xcode, but in other cases, UIKit will throw an exception, terminating the app. An example crash log can be found below.

If all Castle APIs are meant to be called from the main thread only, I think this should be documented (if it isn't already; I might have just missed the relevant docs).

### Workaround
Make sure to only call into Castle on the main thread, for example by checking `Thread.isMainThread` and dispatching to the main queue if necessary.

### Environment
- Castle SDK 2.0.1, installed via CocoaPods
- Seen on both iOS 13 and 14

This was caught during a TestFlight beta phase, so we've only seen 4 crashes so far.

### Crash report
```
Hardware Model:     iPhone12,1
Role:               Foreground
OS Version:         iOS 14.5.1


NSInternalInconsistencyException: threading violation: expected the main thread

0  CoreFoundation          ___exceptionPreprocess
1  libobjc.A.dylib         _objc_exception_throw
2  CoreFoundation          +[NSException raise:format:arguments:]
3  Foundation              -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:]
4  FrontBoardServices      -[FBSSerialQueue assertBarrierOnQueue]
5  FrontBoardServices      -[FBSScene updateClientSettings:withTransitionContext:]
6  FrontBoardServices      -[FBSScene updateClientSettingsWithTransitionBlock:]
7  FrontBoardServices      -[FBSScene updateClientSettingsWithBlock:]
8  UIKitCore               -[FBSScene(UIApp) updateUIClientSettingsWithBlock:]
9  UIKitCore               -[UIApplication setProximityEventsEnabled:]
10 UIKitCore               -[UIDevice setProximityMonitoringEnabled:]
11 <redacted app>          static Hardware.hasProximityMonitoring() (Hardware.swift:197:45)
12 <compiler-generated>    static Hardware.flags()
13 <redacted app>          Generator.generate() (Generator.swift:58:28)
14 <redacted app>          Highwind.token() (Highwind.swift:43:24)
15 <redacted app>          +[Castle clientId] (Castle.m:510:12)
16 <redacted app>          +[CASContext snapshotContext] (CASContext.m:23:24)
17 <redacted app>          -[CASEvent context] (CASEvent.m:134:20)
18 <redacted app>          -[CASEvent JSONPayload] (CASEvent.m:109:56)
19 Foundation              -[NSObject(NSKeyValueCoding) valueForKey:]
20 Foundation              -[NSArray(NSKeyValueCoding) valueForKey:]
21 <redacted app>          -[CASBatch JSONPayload] (CASBatch.m:42:25)
22 <redacted app>          -[CASModel JSONData] (CASModel.m:22:18)
23 <redacted app>          +[Castle flush] (Castle.m:322:69)
24 <redacted app>          +[Castle reset] (Castle.m:363:5)
```

**Comment by aalekz** [comment]
Hey @tiwoc!

Thanks for reporting this issue and for the thorough description (as usual).

The APIs have been intended to work on any thread but the new fingerprinting code, as you pointed out, is assuming that it's called on the main thread.

I will add a workaround so that the Highwind library is called on the main thread. We're planning on releasing 2.0.2 in the coming days so the fix will be included in that release.

**Comment by tiwoc** [comment]
Thanks!

**Comment by aalekz** [comment]
Hi again @tiwoc,

I just released version 2.1.0 that contains a fix for this issue.
